lib/HLSL/DxilValidation.cpp

```diff
@@ -1131,6 +1131,112 @@ static int GetCBufSize(Value *cbHandle, ValidationContext &ValCtx) {
   return RP.CBufferSizeInBytes;
 }
 
+// Make sure none of the handle arguments are undef / zero-initializer,
+// Also, do not accept any resource handles with invalid dxil resource
+// properties
+void ValidateHandleArgsForInstruction(CallInst *CI, DXIL::OpCode opcode,
+                                      ValidationContext &ValCtx) {
+
+  for (Value *op : CI->operands()) {
+    const Type *pHandleTy = ValCtx.HandleTy; // This is a resource handle
+    const Type *pNodeHandleTy = ValCtx.DxilMod.GetOP()->GetNodeHandleType();
+    const Type *pNodeRecordHandleTy =
+        ValCtx.DxilMod.GetOP()->GetNodeRecordHandleType();
+
+    const Type *argTy = op->getType();
+    if (argTy == pNodeHandleTy || argTy == pNodeRecordHandleTy ||
+        argTy == pHandleTy) {
+      bool bIsUninitialized =
+          isa<UndefValue>(op) || isa<ConstantAggregateZero>(op);
+      if (argTy == pHandleTy) {
+        hlsl::DxilResourceProperties DRP = GetResourceFromHandle(op, ValCtx);
+        if (!DRP.isValid()) {
+          ValCtx.EmitInstrError(CI, ValidationRule::InstrResourceInvalid);
+        }
+      }
+      if (bIsUninitialized) {
+        ValCtx.EmitInstrError(CI, ValidationRule::InstrNoReadingUninitialized);
+      }
+    }
+  }
+}
+
+void ValidateHandleArgs(CallInst* CI, DXIL::OpCode opcode, ValidationContext &ValCtx) {
+
+  switch (opcode) {
```

Comment:
My previous comment regarding the switch isn't visible in the file after changes.  So I'm putting this comment here.

My point was that we shouldn't be listing all the ops with handle args we want to check in the switch statement. We should only detect and skip the generic check on the specific known exceptions.  All other ops should just get the generic check. The list of ops that have handle args is an easy way to miss one or to forget to add one in the future. All ops should run the generic handle validation, except several specific handle initialization operations.